### PR TITLE
add unit test for ResourceBindingController

### PR DESCRIPTION
**Pull request by yy158775**
Signed-off-by: yy158775 <1584616775@qq.com>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Add unit tests for binding_contoller using fake client.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controllers/binding/binding_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+	return bindController
+}
+
+func newFixture(t *testing.T) *fixture {
+	f := &fixture{}
+	f.t = t
+	return f
+}
+
+// 只针对deployment 还是说所有
```
Comment (RainbowMango) [inline pkg/controllers/binding/binding_controller_test.go:290]: Don't comment in Chinese...

**Comment by XiShanYongYe-Chang** [comment]
Some additions and modifications are still needed.
/hold

File: pkg/controllers/binding/binding_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+	f.resourceBindings = append(f.resourceBindings, &rb)
+	f.clusters = append(f.clusters, &c[0], &c[1], &c[2])
+
+	f.expectedWorks = append(f.expectedWorks, w...)
+
+	f.objectResourceInit()
+
+	// Wait for informerManager's cache sync
+	_ = f.informerManager.Lister(appsv1.SchemeGroupVersion.WithResource("deployments"))
+	f.informerManager.Start()
```
Comment (XiShanYongYe-Chang) [inline pkg/controllers/binding/binding_controller_test.go:410]: Does this step is needed?

The fakeClient may not be suitable for the informer: https://github.com/kubernetes/client-go/blob/c81636cd4415f0858a99c2cc18023740df7aa15f/examples/fake-client/main_test.go#L78-L81

File: pkg/controllers/binding/binding_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+		}
+
+		_, err = f.dynamicClient.Resource(clusterGVR).Namespace(c.Namespace).Create(context.TODO(), unstructured, metav1.CreateOptions{})
+		if err != nil {
+			f.t.Fatal(err)
+		}
+	}
+}
+
+// Split the objectResourceInit in order to pass the CI
```
[the remark this answers is not in the document]
Comment (yy158775) [reply pkg/controllers/binding/binding_controller_test.go:234]: If I combine it into objectResourceInit,the objectResourceInit can't pass the CI and the error is  "cyclomatic complexity is high".

File: pkg/controllers/binding/binding_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,592 @@
...
+		DynamicClient:       f.dynamicClient,
+		InformerManager:     f.informerManager,
+		EventRecorder:       f.eventRecorder,
+		RESTMapper:          f.restMapper,
+		OverrideManager:     f.overrideManager,
+		ResourceInterpreter: f.resourceInterpreter,
+		RateLimiterOptions:  ratelimiterflag.Options{},
+	}
+
+	// not return
```
[the remark this answers is not in the document]
Comment (yy158775) [reply pkg/controllers/binding/binding_controller_test.go:280]: The goroutine will be blocked and doesn't exit.

File: pkg/controllers/binding/binding_controller_test.go (hunk 2 above)
Comment (yy158775) [reply pkg/controllers/binding/binding_controller_test.go:410]: This step is needed for informerManager to sync deployments object.The informanager will not have the "deployments" informer and object store without this step. 
According to https://github.com/kubernetes/client-go/blob/c81636cd4415f0858a99c2cc18023740df7aa15f/examples/fake-client/main_test.go#L74-L77,it doesn't support resource version,so we can add objects to fake client before the informer created.Then the informer will use List method to get all objects.

**Comment by RainbowMango** [comment]
@yy158775 Please rebase your code and let your commit based on the latest code on the master branch.
```
$ git rebase master
$ git push origin rbcontroller_test -f 
```

**Comment by RainbowMango** [comment]
I'm afraid the unit test is more complex than the code it protects.
Let's postpone it and try to find a cheaper way.
/close

**Pull request closed without merging**